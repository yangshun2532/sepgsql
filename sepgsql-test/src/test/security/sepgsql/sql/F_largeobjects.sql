--
-- SE-PostgreSQL testcases : Largeobjects
--
-- selinux_user: unconfined_u
-- selinux_role: unconfined_r
-- selinux_type: sepgsql_test_t
-- selinux_range: s0:c0

SELECT lo_get_security(6001);
SELECT lo_get_security(6002);
SELECT lo_get_security(6003);

SELECT security_context, loid, COUNT(*) FROM pg_largeobject GROUP BY security_context, loid;

-- read large object
BEGIN;
SELECT lo_open(6001, 4 * 65536);
SELECT loread(0, 24);
ROLLBACK;

BEGIN;
SELECT lo_open(6002, 4 * 65536);
SELECT loread(0, 24);
ROLLBACK;

BEGIN;
SELECT lo_open(6003, 4 * 65536);
SELECT loread(0, 24);					-- to be failed
ROLLBACK;

-- write large object
BEGIN;
SELECT lo_open(6001, 2 * 65536);
SELECT lowrite(0, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ');
COMMIT;

BEGIN;
SELECT lo_open(6002, 2 * 65536);
SELECT lowrite(0, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ');	-- to be failed
COMMIT;

BEGIN;
SELECT lo_open(6003, 2 * 65536);
SELECT lowrite(0, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ');	-- to be failed
COMMIT;

-- create large object
BEGIN;
SELECT lo_create(6004);
SELECT lo_open(6004, 2 * 65536);
SELECT lowrite(0, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ');
SELECT lo_close(0);
COMMIT;

-- append large object
BEGIN;
SELECT lo_open(6001, 6 * 65536);
SELECT lo_lseek(0, 0, 2);	-- seek to end
SELECT lowrite(0, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ');
SELECT lo_close(0);
COMMIT;

-- confirm results
SELECT security_context, loid, COUNT(*) FROM pg_largeobject GROUP BY security_context, loid;
