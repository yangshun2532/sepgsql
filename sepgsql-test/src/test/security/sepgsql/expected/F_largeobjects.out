--
-- SE-PostgreSQL testcases : Largeobjects
--
-- selinux_user: unconfined_u
-- selinux_role: unconfined_r
-- selinux_type: sepgsql_test_t
-- selinux_range: s0:c0
SELECT lo_get_security(6001);
           lo_get_security           
-------------------------------------
 system_u:object_r:sepgsql_blob_t:s0
(1 row)

SELECT lo_get_security(6002);
            lo_get_security             
----------------------------------------
 system_u:object_r:sepgsql_ro_blob_t:s0
(1 row)

SELECT lo_get_security(6003);
              lo_get_security               
--------------------------------------------
 system_u:object_r:sepgsql_secret_blob_t:s0
(1 row)

SELECT security_context, loid, COUNT(*) FROM pg_largeobject GROUP BY security_context, loid;
              security_context              | loid | count 
--------------------------------------------+------+-------
 system_u:object_r:sepgsql_ro_blob_t:s0     | 6002 |    18
 system_u:object_r:sepgsql_secret_blob_t:s0 | 6003 |    18
 system_u:object_r:sepgsql_blob_t:s0        | 6001 |    18
(3 rows)

-- read large object
BEGIN;
SELECT lo_open(6001, 4 * 65536);
 lo_open 
---------
       0
(1 row)

SELECT loread(0, 24);
                                              loread                                              
--------------------------------------------------------------------------------------------------
 \000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000
(1 row)

ROLLBACK;
BEGIN;
SELECT lo_open(6002, 4 * 65536);
 lo_open 
---------
       0
(1 row)

SELECT loread(0, 24);
                                              loread                                              
--------------------------------------------------------------------------------------------------
 \000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000
(1 row)

ROLLBACK;
BEGIN;
SELECT lo_open(6003, 4 * 65536);
 lo_open 
---------
       0
(1 row)

SELECT loread(0, 24);					-- to be failed
psql:./F_largeobjects.sql:28: ERROR:  SELinux: denied { read } scontext=unconfined_u:unconfined_r:sepgsql_test_t:s0:c0 tcontext=system_u:object_r:sepgsql_secret_blob_t:s0 tclass=db_blob name=loid:6003
ROLLBACK;
-- write large object
BEGIN;
SELECT lo_open(6001, 2 * 65536);
 lo_open 
---------
       0
(1 row)

SELECT lowrite(0, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ');
 lowrite 
---------
      26
(1 row)

COMMIT;
BEGIN;
SELECT lo_open(6002, 2 * 65536);
 lo_open 
---------
       0
(1 row)

SELECT lowrite(0, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ');	-- to be failed
psql:./F_largeobjects.sql:39: ERROR:  SELinux: denied { write } scontext=unconfined_u:unconfined_r:sepgsql_test_t:s0:c0 tcontext=system_u:object_r:sepgsql_ro_blob_t:s0 tclass=db_blob name=loid:6002
COMMIT;
BEGIN;
SELECT lo_open(6003, 2 * 65536);
 lo_open 
---------
       0
(1 row)

SELECT lowrite(0, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ');	-- to be failed
psql:./F_largeobjects.sql:44: ERROR:  SELinux: denied { write } scontext=unconfined_u:unconfined_r:sepgsql_test_t:s0:c0 tcontext=system_u:object_r:sepgsql_secret_blob_t:s0 tclass=db_blob name=loid:6003
COMMIT;
-- create large object
BEGIN;
SELECT lo_create(6004);
 lo_create 
-----------
      6004
(1 row)

SELECT lo_open(6004, 2 * 65536);
 lo_open 
---------
       0
(1 row)

SELECT lowrite(0, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ');
 lowrite 
---------
      26
(1 row)

SELECT lo_close(0);
 lo_close 
----------
        0
(1 row)

COMMIT;
-- append large object
BEGIN;
SELECT lo_open(6001, 6 * 65536);
 lo_open 
---------
       0
(1 row)

SELECT lo_lseek(0, 0, 2);	-- seek to end
 lo_lseek 
----------
    36864
(1 row)

SELECT lowrite(0, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ');
 lowrite 
---------
      26
(1 row)

SELECT lo_close(0);
 lo_close 
----------
        0
(1 row)

COMMIT;
-- confirm results
SELECT security_context, loid, COUNT(*) FROM pg_largeobject GROUP BY security_context, loid;
              security_context              | loid | count 
--------------------------------------------+------+-------
 system_u:object_r:sepgsql_ro_blob_t:s0     | 6002 |    18
 system_u:object_r:sepgsql_secret_blob_t:s0 | 6003 |    18
 unconfined_u:object_r:sepgsql_blob_t:s0:c0 | 6004 |     1
 system_u:object_r:sepgsql_blob_t:s0        | 6001 |    19
(4 rows)

